lib/Optimizer/Transforms/Mapping.cpp

```diff
@@ -0,0 +1,624 @@
+
+  double computeLayerCost(ArrayRef<VirtualOp> layer);
+
+  Swap chooseSwap();
+
+private:
+  const Device &device;
+  WireMap &wireToVirtualQ;
+  Placement &placement;
+
+  // Parameters
+  unsigned extendedLayerSize = 20;
+  float extendedLayerWeight = 0.5;
+  float decayDelta = 0.5;
+  unsigned roundsDecayReset = 5;
+
+  // Internal data
+  SmallVector<VirtualOp> frontLayer;
+  SmallVector<VirtualOp> extendedLayer;
+  SmallVector<VirtualOp> measureLayer;
+  llvm::SmallPtrSet<mlir::Operation *, 32> measureLayerSet;
+  llvm::SmallSet<Placement::DeviceQ, 32> involvedPhy;
+  SmallVector<float> phyDecay;
+
+  SmallVector<Value> phyToWire;
+
+  /// Keeps track of how many times an operation was visited.
+  DenseMap<Operation *, unsigned> visited;
+
+  /// Keep track of whether or not we're in the phase that allows measurements
+  /// to be mapped
+  bool allowMeasurementMapping;
+
+#ifndef NDEBUG
+  /// A logger used to emit diagnostics during the maping process.
+  llvm::ScopedPrinter logger{llvm::dbgs()};
+#endif
+};
+
+void SabreRouter::visitUsers(ResultRange::user_range users,
+                             SmallVectorImpl<VirtualOp> &layer,
+                             SmallVectorImpl<Operation *> *incremented) {
+  for (auto user : users) {
+    auto [entry, created] = visited.try_emplace(user, 1);
+    if (!created)
+      entry->second += 1;
+    if (incremented)
+      incremented->push_back(user);
+
+    if (!quake::isSupportedMappingOperation(user)) {
+      auto tmpOp = dyn_cast<mlir::Operation *>(user);
+      LLVM_DEBUG({
+        logger.getOStream() << "WARNING: unsupported op: " << *tmpOp << '\n';
+      });
+    } else {
+      auto wires = quake::getQuantumOperands(user);
+      if (entry->second == wires.size()) {
+        SmallVector<Placement::VirtualQ, 2> qubits;
+        for (auto wire : wires)
+          qubits.push_back(wireToVirtualQ[wire]);
+        // Don't process measurements until we're ready
+        if (allowMeasurementMapping || !user->hasTrait<QuantumMeasure>()) {
+          layer.emplace_back(user, qubits);
+        } else {
+          // Add to measureLayer. Don't add duplicates.
+          if (measureLayerSet.find(user) == measureLayerSet.end()) {
+            measureLayer.emplace_back(user, qubits);
+            measureLayerSet.insert(user);
+          }
+        }
+      }
+    }
+  }
+}
+
+LogicalResult SabreRouter::mapOperation(VirtualOp &virtOp) {
+  // Take the device qubits from this operation.
+  SmallVector<Placement::DeviceQ, 2> deviceQubits;
+  for (auto vr : virtOp.qubits)
+    deviceQubits.push_back(placement.getPhy(vr));
+
+  // An operation cannot be mapped if it is not a measurement and uses two
+  // qubits virtual qubit that are no adjacently placed.
+  if (!virtOp.op->hasTrait<QuantumMeasure>() && deviceQubits.size() == 2 &&
+      !device.areConnected(deviceQubits[0], deviceQubits[1]))
+    return failure();
+
+  // Rewire the operation.
+  SmallVector<Value, 2> newOpWires;
```

Comment:
> Is `2` needed? Better to just let it default unless we know this can never be `> 2` in length.

Yes, 2 is the desired value here. Leaving it out would likely make compiler to reserve additional space that is not needed. The check in line 510 verifies that and fails the pass if that is not the case.